src/v/config/constraint_store.h

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "config/configuration.h"
+#include "config/constraint_types.h"
+
+#include <absl/container/flat_hash_map.h>
+
+namespace config {
+/**
+ * A store to track configuration constraints. This assists with de-coupling
+ * constraint arguments from their methods that are used to validate or clamp
+ * topic properties. Additionally, adding and searching for constraints is
+ * easier.
+ */
+class constraint_store {
```

Comment:
This could be replaced with several helper methods for searching and inserting constraints. Created a follow-up ticket here https://github.com/redpanda-data/redpanda/issues/14774